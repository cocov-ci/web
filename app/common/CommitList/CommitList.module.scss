@use "sass:map";
@import "styles/variables.module";

$issuesWidth: 90px;

.base {
  height: 100%;
  width: 100%;
  display: flex;
  flex-direction: column;

  .header {
    display: flex;
    flex-shrink: 1;
    border-bottom: 1px solid $colorGray;
    font-size: 15px;
    color: $colorBlack;
    letter-spacing: 0;
    font-variation-settings: "wgth" 500;

    .commitColumn {
      flex: 1;
      border-right: 1px solid $colorGray;
      padding: 9px 0 9px 32px;
    }

    .countersColumns {
      display: flex;
      width: 200px;
      text-align: center;

      > div {
        padding: 9px;
      }

      .issuesColumn {
        width: $issuesWidth;
        border-right: 1px solid $colorGray;
      }

      .coverageColumn {
        flex: 1;
      }
    }
  }

  .rows {
    flex: 1;

    .row {
      position: relative;

      .rowWrapper {
        position: relative;
        display: flex;
        z-index: 1;
        border: 1px solid transparent;
        cursor: pointer;
        transition: background-color 0.1s ease-in-out, border 0.1s ease-in-out;

        &:hover {
          background-color: map.get($colorListItemActive, "background");
          border: 1px solid map.get($colorListItemActive, "border");
          border-radius: 4px;
        }

        .headerWrapper {
          padding-left: 25px;
          padding-top: 8px;
          padding-bottom: 8px;
          flex: 1;
        }

        .counters {
          display: flex;
          width: 200px;
          font-variation-settings: "wght" 600;

          div {
            display: flex;
            justify-content: center;
            align-items: center;
          }

          .issuesCounter {
            width: $issuesWidth;
          }

          .coverageCounter {
            flex: 1;
          }
        }
      }

      &.loading {
        .rowWrapper {
          cursor: default;

          &:hover {
            background-color: transparent;
            border: 1px solid transparent;
          }
        }
      }

      &::after {
        display: block;
        content: " ";
        width: 100%;
        height: 1px;
      }

      &:first-of-type {
        &::before {
          display: block;
          content: " ";
          width: 100%;
          height: 5px;
        }
      }

      &:last-of-type {
        &::after {
          display: block;
          content: " ";
          width: 100%;
          height: 5px;
        }
      }

      .rail {
        width: 4px;
        height: 100%;
        position: absolute;
        background: #e4e4e4;
        left: 36px;
        top: 0;
      }

      .virtualSpacers {
        position: absolute;
        height: 100%;
        width: 100%;
        top: 0;
        z-index: 0;
        padding-left: 1px;
        padding-right: 1px;
        display: flex;

        .commitSpacer {
          flex: 1;
        }

        .countersSpacer {
          display: flex;
          width: 200px;

          div {
            border-left: 1px solid $colorGray;
          }

          .issuesSpacer { width: $issuesWidth; }
          .coverageSpacer { flex: 1; }
        }
      }
    }
  }

  .paging {
    border-top: 1px solid $colorGray;

    .paginationComponent {
      padding: 10px 0;
      margin: 0;
    }
  }
}
