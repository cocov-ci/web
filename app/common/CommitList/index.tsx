'use client'

import classNames from 'classnames'
import Link from 'next/link'
import { useSearchParams } from 'next/navigation'
import React from 'react'

import CommitHeader from 'app/common/CommitHeader'
import Loading from 'app/common/CommitHeader/Loading'
import Pagination from 'app/common/Pagination'
import { CommitHeaderProps } from 'types/Branches'
import Url from 'types/Url'

import styles from './CommitList.module.scss'

export type CommitListItem = {
  issues?: number
  coverage?: number
  commitInfo: CommitHeaderProps
}

type CommitListProps = {
  className?: string
  commits: Array<CommitListItem>
  paginationData?: {
    total: number
    currentPage: number
    onPageClick: () => void
  }
  onChangePage?: (requestedPage: number) => void
  loading?: boolean
}

type RowProps = {
  children?: React.ReactNode
  className?: string
  href: Url
}

type BaseRowProps = {
  children?: React.ReactNode
  className?: string
}

const BaseRow = ({ children, className }: BaseRowProps) => (
  <div className={classNames(styles.row, className)}>
    <div className={styles.rail} />
    <div className={styles.virtualSpacers}>
      <div className={styles.commitSpacer} />
      <div className={styles.countersSpacer}>
        <div className={styles.issuesSpacer} />
        <div className={styles.coverageSpacer} />
      </div>
    </div>
    <div className={styles.rowWrapper}>{children}</div>
  </div>
)

const Row = ({ children, className, href }: RowProps) => (
  <Link href={href}>
    <BaseRow className={className}>{children}</BaseRow>
  </Link>
)

const CommitList = ({
  className,
  commits,
  paginationData,
  loading,
  onChangePage,
}: CommitListProps) => {
  const searchParams = useSearchParams()
  const repositoryName = searchParams.get('repositoryName')
  const branchName = searchParams.get('branchName')

  const commitList = () =>
    commits.map(c => (
      <Row
        href={`/repos/${repositoryName}/branches/${branchName}/commits/${c.commitInfo.headSHA}`}
        key={c.commitInfo.headSHA}
      >
        <div className={styles.headerWrapper}>
          <CommitHeader {...c.commitInfo} />
        </div>
        <div className={styles.counters}>
          <div className={styles.issuesCounter}>{c.issues}</div>
          <div className={styles.coverageCounter}>{c.coverage}%</div>
        </div>
      </Row>
    ))

  const loadingList = () =>
    new Array(5).fill(0).map((e, idx) => (
      // eslint-disable-next-line react/no-array-index-key
      <BaseRow className={styles.loading} key={`loading_${idx}`}>
        <div className={styles.headerWrapper}>
          <Loading />
        </div>
        <div className={styles.counters}>
          <div className={styles.issuesCounter} />
          <div className={styles.coverageCounter} />
        </div>
      </BaseRow>
    ))

  return (
    <div className={classNames(styles.base, className)}>
      <div className={styles.header}>
        <div className={styles.commitColumn}>Commit</div>
        <div className={styles.countersColumns}>
          <div className={styles.issuesColumn}>Issues</div>
          <div className={styles.coverageColumn}>Coverage</div>
        </div>
      </div>

      <div className={styles.rows}>
        {!loading ? commitList() : loadingList()}
      </div>

      <div className={styles.paging}>
        {paginationData && (
          <Pagination
            {...paginationData}
            className={styles.paginationComponent}
            onPageClick={p => onChangePage && onChangePage(p)}
          />
        )}
      </div>
    </div>
  )
}

export default CommitList
