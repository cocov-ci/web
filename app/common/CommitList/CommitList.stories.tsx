import type { Meta, StoryObj } from '@storybook/react'

import CommitList from './index'

const meta: Meta<typeof CommitList> = {
  title: 'Common/CommitList',
  component: CommitList,
  tags: ['autodocs'],
}

export default meta

type Story = StoryObj<typeof CommitList>

export const Default: Story = {
  args: {
    paginationData: {
      total: 10,
      currentPage: 2,
    },
    commits: [
      {
        commitInfo: {
          headURL: 'https://github.com/cocov-ci/web',
          avatarURL:
            'https://avatars.githubusercontent.com/u/118852412?s=200&v=4',
          headSHA: '62bad0cbadbeefc0c0f',
          commitMessage:
            'fix(rollout-restart): Ensure to keep watching rollout status',
          username: 'cocov-ci',
          timestamp: new Date(Date.parse('2022-11-22T21:25:31Z')),
          readonly: true,
        },
        coverage: 100,
        issues: 9999,
      },
      {
        commitInfo: {
          headURL: 'https://github.com/cocov-ci/web',
          avatarURL:
            'https://avatars.githubusercontent.com/u/118852412?s=200&v=4',
          headSHA: '62bad0cbadbeefc0c0f1',
          commitMessage:
            'fix(rollout-restart): Ensure to keep watching rollout status',
          username: 'cocov-ci',
          timestamp: new Date(Date.parse('2022-11-22T21:25:31Z')),
          readonly: true,
        },
        coverage: 100,
        issues: 9999,
      },
      {
        commitInfo: {
          headURL: 'https://github.com/cocov-ci/web',
          avatarURL:
            'https://avatars.githubusercontent.com/u/118852412?s=200&v=4',
          headSHA: '62bad0cbadbeefc0c0f2',
          commitMessage:
            'fix(rollout-restart): Ensure to keep watching rollout status',
          username: 'cocov-ci',
          timestamp: new Date(Date.parse('2022-11-22T21:25:31Z')),
          readonly: true,
        },
        coverage: 100,
        issues: 9999,
      },
    ],
  },
}
