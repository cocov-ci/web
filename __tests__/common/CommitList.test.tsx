import { fireEvent, render, screen } from '@testing-library/react'

import CommitList from 'app/common/CommitList'

import '@testing-library/jest-dom'

jest.mock('next/navigation', () => ({
  useSearchParams: jest.fn(() => ({
    get: jest.fn(),
  })),
}))

const args = {
  paginationData: {
    total: 10,
    currentPage: 2,
    onPageClick: () => null,
  },
  commits: [
    {
      commitInfo: {
        headURL: 'https://github.com/cocov-ci/web',
        avatarURL:
          'https://avatars.githubusercontent.com/u/118852412?s=200&v=4',
        headSHA: '62bad0cbadbeefc0c0f',
        commitMessage:
          'fix(rollout-restart): Ensure to keep watching rollout status',
        username: 'cocov-ci',
        timestamp: new Date(Date.parse('2022-11-22T21:25:31Z')),
      },
      coverage: 100,
      issues: 9999,
    },
    {
      commitInfo: {
        headURL: 'https://github.com/cocov-ci/web',
        avatarURL:
          'https://avatars.githubusercontent.com/u/118852412?s=200&v=4',
        headSHA: '62bad0cbadbeefc0c0f1',
        commitMessage:
          'fix(rollout-restart): Ensure to keep watching rollout status',
        username: 'cocov-ci',
        timestamp: new Date(Date.parse('2022-11-22T21:25:31Z')),
      },
      coverage: 100,
      issues: 9999,
    },
  ],
}

describe('common/CommitList', () => {
  it('renders CommitList', () => {
    const { container } = render(<CommitList {...args} />)

    expect(container.querySelectorAll('.rows > a').length).toEqual(2)
  })

  it('renders CommitList with "loading" param', () => {
    const { container } = render(<CommitList {...args} loading={true} />)

    expect(container.querySelectorAll('.rows > a').length).toEqual(0)
    expect(container.querySelectorAll('.loading.row').length).toEqual(5)
  })

  it('triggers onChangePage event', () => {
    const handleEvent = jest.fn()

    const { container } = render(
      <CommitList {...args} onChangePage={handleEvent} />,
    )
    fireEvent.click(container.querySelectorAll('li.page')[0])
    expect(handleEvent).toHaveBeenCalledTimes(1)
  })

  it('renders CommitList snapshots', () => {
    const { container } = render(<CommitList {...args} />)

    expect(container).toMatchSnapshot()
  })
})
