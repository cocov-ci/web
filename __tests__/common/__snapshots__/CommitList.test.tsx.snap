// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common/CommitList renders CommitList snapshots 1`] = `
<div>
  <div
    class="base"
  >
    <div
      class="header"
    >
      <div
        class="commitColumn"
      >
        Commit
      </div>
      <div
        class="countersColumns"
      >
        <div
          class="issuesColumn"
        >
          Issues
        </div>
        <div
          class="coverageColumn"
        >
          Coverage
        </div>
      </div>
    </div>
    <div
      class="rows"
    >
      <a
        href="/repos/undefined/branches/undefined/commits/62bad0cbadbeefc0c0f"
      >
        <div
          class="row"
        >
          <div
            class="rail"
          />
          <div
            class="virtualSpacers"
          >
            <div
              class="commitSpacer"
            />
            <div
              class="countersSpacer"
            >
              <div
                class="issuesSpacer"
              />
              <div
                class="coverageSpacer"
              />
            </div>
          </div>
          <div
            class="rowWrapper"
          >
            <div
              class="headerWrapper"
            >
              <div
                class="base"
              >
                <div
                  class="avatarWrapper"
                >
                  <img
                    alt=""
                    aria-hidden="true"
                    class="base"
                    src="https://avatars.githubusercontent.com/u/118852412?s=200&v=4"
                    style="height: 25px; width: 25px;"
                  />
                </div>
                <div
                  class="infoWrapper"
                >
                  <div
                    class="commitMeta"
                  >
                    <span
                      class="username"
                    >
                      cocov-ci
                    </span>
                    <span>
                      committed
                    </span>
                    <a
                      class="headLink className"
                      href="https://github.com/cocov-ci/web"
                      rel="noreferrer"
                      target="_blank"
                    >
                      62bad0
                    </a>
                    <span
                      class=""
                      title="Nov 22, 2022, 9:25 PM GMT"
                    >
                      4 months ago
                    </span>
                  </div>
                  <div
                    class="commitTitle"
                  >
                    <a
                      href="https://github.com/cocov-ci/web"
                      rel="noreferrer"
                      target="_blank"
                    >
                      fix(rollout-restart): Ensure to keep watching rollout status
                    </a>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="counters"
            >
              <div
                class="issuesCounter"
              >
                9999
              </div>
              <div
                class="coverageCounter"
              >
                100
                %
              </div>
            </div>
          </div>
        </div>
      </a>
      <a
        href="/repos/undefined/branches/undefined/commits/62bad0cbadbeefc0c0f1"
      >
        <div
          class="row"
        >
          <div
            class="rail"
          />
          <div
            class="virtualSpacers"
          >
            <div
              class="commitSpacer"
            />
            <div
              class="countersSpacer"
            >
              <div
                class="issuesSpacer"
              />
              <div
                class="coverageSpacer"
              />
            </div>
          </div>
          <div
            class="rowWrapper"
          >
            <div
              class="headerWrapper"
            >
              <div
                class="base"
              >
                <div
                  class="avatarWrapper"
                >
                  <img
                    alt=""
                    aria-hidden="true"
                    class="base"
                    src="https://avatars.githubusercontent.com/u/118852412?s=200&v=4"
                    style="height: 25px; width: 25px;"
                  />
                </div>
                <div
                  class="infoWrapper"
                >
                  <div
                    class="commitMeta"
                  >
                    <span
                      class="username"
                    >
                      cocov-ci
                    </span>
                    <span>
                      committed
                    </span>
                    <a
                      class="headLink className"
                      href="https://github.com/cocov-ci/web"
                      rel="noreferrer"
                      target="_blank"
                    >
                      62bad0
                    </a>
                    <span
                      class=""
                      title="Nov 22, 2022, 9:25 PM GMT"
                    >
                      4 months ago
                    </span>
                  </div>
                  <div
                    class="commitTitle"
                  >
                    <a
                      href="https://github.com/cocov-ci/web"
                      rel="noreferrer"
                      target="_blank"
                    >
                      fix(rollout-restart): Ensure to keep watching rollout status
                    </a>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="counters"
            >
              <div
                class="issuesCounter"
              >
                9999
              </div>
              <div
                class="coverageCounter"
              >
                100
                %
              </div>
            </div>
          </div>
        </div>
      </a>
    </div>
    <div
      class="paging"
    >
      <div
        class="pagination paginationComponent"
      >
        <ul
          class="pages"
        >
          <svg
            class="arrow"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="17 18 11 12 17 6"
            />
            <path
              d="M7 6v12"
            />
          </svg>
          <li
            class="page"
          >
            1
          </li>
          <li
            class="page active"
          >
            2
          </li>
          <li
            class="page"
          >
            3
          </li>
          <li
            class="page"
          >
            4
          </li>
          <li
            class="page"
          >
            5
          </li>
          <li
            class="page"
          >
            6
          </li>
          <li
            class="page"
          >
            7
          </li>
          <li
            class="page"
          >
            8
          </li>
          <li
            class="page"
          >
            9
          </li>
          <li
            class="page"
          >
            10
          </li>
          <svg
            class="arrow"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="7 18 13 12 7 6"
            />
            <path
              d="M17 6v12"
            />
          </svg>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
